kernel/kernel-api/src/main/java/io/delta/kernel/internal/TableImpl.java

```diff
@@ -164,4 +167,60 @@ public CloseableIterator<ColumnarBatch> getChangesByVersion(
 
     return DeltaLogActionUtils.readCommitFiles(engine, commitFiles, readSchema);
   }
+
+  /**
+   * Returns the raw delta actions for each version between startVersion and endVersion. Only reads
+   * the actions requested in actionSet from the JSON log files. When {@link
+   * DeltaLogActionUtils.DeltaAction#PROTOCOL} is present in {@code actionSet} Kernel also performs
+   * read protocol checks on any protocol actions read.
+   *
+   * <p>For the returned columnar batches:
+   *
+   * <ul>
+   *   <li>Each row within the same batch is guaranteed to have the same commit version
+   *   <li>The batch commit versions are monotonically increasing
+   *   <li>The top-level columns include "version", "timestamp", and the actions requested in
+   *       actionSet. "version" and "timestamp" are the first and second columns in the schema,
+   *       respectively. The remaining columns are based on the actions requested and each have the
+   *       schema found in {@code DeltaAction.schema}.
+   * </ul>
+   *
+   * @param engine {@link Engine} instance to use in Delta Kernel.
+   * @param startVersion start version (inclusive)
+   * @param endVersion end version (inclusive)
+   * @param actionSet the actions to read and return from the JSON log files
+   * @return an iterator of batches where each row in the batch has exactly one non-null action and
+   *     its commit version and timestamp
+   * @throws TableNotFoundException if the table does not exist or if it is not a delta table
+   * @throws KernelException if a commit file does not exist for any of the versions in the provided
+   *     range
+   * @throws KernelException if provided an invalid version range
+   * @throws KernelException if {@link DeltaLogActionUtils.DeltaAction#PROTOCOL} is in actionSet and
+   *     a Protocol action that is not read supported by Kernel is read
+   */
+  public CloseableIterator<ColumnarBatch> getChanges(
+      Engine engine,
+      long startVersion,
+      long endVersion,
+      Set<DeltaLogActionUtils.DeltaAction> actionSet) {
+    if (actionSet.contains(DeltaLogActionUtils.DeltaAction.PROTOCOL)) {
+      return getRawChanges(engine, startVersion, endVersion, actionSet)
+          .map(
+              batch -> {
+                int protocolIdx =
```

Comment:
get the protocol col index once outside of the `.map` on line 207.

Reply:
We can't access the schema w/o the batch